File: expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt
package expo.modules.maps

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.uimanager.UIManagerModule
import com.google.android.gms.maps.model.LatLng
import expo.modules.core.interfaces.services.UIManager
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.maps.googleMaps.GoogleMapsView
import expo.modules.maps.googleMaps.events.GoogleMapsEventEmitterManager
import expo.modules.maps.googleMaps.events.MapEventsNames

class ExpoGoogleMapsModule : Module() {

  override fun definition() = ModuleDefinition {
    Name("ExpoGoogleMaps")

    Events(events = MapEventsNames.values().map { it.eventName }.toTypedArray())
    println(MapEventsNames.values().map { it.eventName }.toTypedArray().contentDeepToString())

    val googleMapsEventEmitterManager = GoogleMapsEventEmitterManager(::sendEvent)
    
    AsyncFunction("getSearchCompletions") { viewHandle: Int, searchQueryFragment: String, promise: Promise ->
      val rnContext = appContext.reactContext as? ReactApplicationContext ?: return@AsyncFunction
      val uiManager = rnContext.getNativeModule(UIManagerModule::class.java) ?: return@AsyncFunction
      appContext.activityProvider?.currentActivity?.runOnUiThread {
        val view = uiManager.resolveView(viewHandle) as GoogleMapsView
        view.fetchPlacesSearchCompletions(searchQueryFragment, promise)
      }
    }

    ViewManager {
      Events("onMapClick","onMapReady", "onMapLoaded", "onRegionChange",
        "onRegionChangeComplete","onRegionChangeStarted", "onPoiClick")

      View {
        GoogleMapsView(it).also { googleMapsView ->
          appContext.legacyModule<UIManager>()
            ?.registerLifecycleEventListener(googleMapsView.lifecycleEventListener)
          googleMapsView.registerEvents(googleMapsEventEmitterManager)
        }
      }


      Prop("enableRotateGestures") { view: GoogleMapsView, enable: Boolean ->
        view.setEnabledRotateGestures(enable)
      }

      Prop("enableScrollGestures") { view: GoogleMapsView, enable: Boolean ->
        view.setEnabledScrollGestures(enable)
      }

      Prop("enableTiltRotateGestures") { view: GoogleMapsView, enable: Boolean ->
        view.setEnabledTiltGestures(enable)
      }

      Prop("enableZoomGestures") { view: GoogleMapsView, enable: Boolean ->
        view.setEnabledZoomGestures(enable)
      }

      Prop("mapType") { view: GoogleMapsView, mapType: MapType ->
        view.setMapType(mapType)
      }

      Prop("showZoomControls") { view: GoogleMapsView, enable: Boolean ->
        view.setShowZoomControl(enable)
      }

      Prop("showCompass") { view: GoogleMapsView, enable: Boolean ->
        view.setShowCompass(enable)
      }

      Prop("showMapToolbar") { view: GoogleMapsView, enable: Boolean ->
        view.setShowMapToolbar(enable)
      }

      Prop("showMyLocationButton") { view: GoogleMapsView, enable: Boolean ->
        view.setShowMyLocationButton(enable)
      }

      Prop("showLevelPicker") { view: GoogleMapsView, enable: Boolean ->
        view.setShowLevelPicker(enable)
      }

      Prop("googleMapsJsonStyleString") { view: GoogleMapsView, jsonStyleString: String ->
        view.setMapStyle(jsonStyleString)
      }

      Prop("markers") { view: GoogleMapsView, markerObjects: Array<MarkerObject> ->
        view.setMarkers(markerObjects)
      }

      Prop("polygons") { view: GoogleMapsView, polygonObjects: Array<PolygonObject> ->
        view.setPolygons(polygonObjects)
      }

      Prop("polylines") { view: GoogleMapsView, polylineObjects: Array<PolylineObject> ->
        view.setPolylines(polylineObjects)
      }

      Prop("initialCameraPosition") { view: GoogleMapsView, initialCameraPosition: CameraPosition ->
        view.setInitialCameraPosition(initialCameraPosition)
      }

      Prop("circles") { view: GoogleMapsView, circleObjects: Array<CircleObject> ->
        view.setCircles(circleObjects)
      }

      Prop("clusters") { view: GoogleMapsView, clusterObjects: Array<ClusterObject> ->
        view.setClusters(clusterObjects)
      }

      Prop("enableTraffic") { view: GoogleMapsView, enable: Boolean ->
        view.setEnabledTraffic(enable)
      }

      Prop("kmls") { view: GoogleMapsView, kmlObjects: Array<KMLObject> ->
        view.setKMLs(kmlObjects)
      }

      Prop("geojsons") { view: GoogleMapsView, geoJsonObjects: Array<GeoJsonObject> ->
        view.setGeoJsons(geoJsonObjects)
      }

      Prop("overlays") { view: GoogleMapsView, overlayObjects: Array<OverlayObject> ->
        view.setOverlays(overlayObjects)
      }
      
      Prop("heatmaps") { view: GoogleMapsView, heatmapObjects: Array<HeatmapObject> ->
        view.setHeatmaps(heatmapObjects)
      }
      
      Prop("createPOISearchRequest") { view: GoogleMapsView, place: String ->
        view.createPlaceSearchRequest(place)
      }

      Prop("clickablePOIs") { view: GoogleMapsView, arePOIClickable: Boolean ->
        view.setClickablePOIs(arePOIClickable)
      }
    }
  }
}

File: expo-maps/ios/ExpoMaps/ExpoGoogleMapsModule.swift
import ExpoModulesCore

public class ExpoGoogleMapsModule: Module {

  public func definition() -> ModuleDefinition {
    Name("ExpoGoogleMaps")

    ViewManager {
      View {
        GoogleMapsView()
      }

      Prop("showCompass") { (view: GoogleMapsView, enable: Bool) in
        view.setShowCompass(enable: enable)
      }

      Prop("showMyLocationButton") { (view: GoogleMapsView, enable: Bool) in
        view.setShowMyLocationButton(enable: enable)
      }

      Prop("showLevelPicker") { (view: GoogleMapsView, enable: Bool) in
        view.setShowLevelPicker(enable: enable)
      }

      Prop("enableRotateGestures") { (view: GoogleMapsView, enable: Bool) in
        view.setEnabledRotateGestures(enabled: enable)
      }

      Prop("enableScrollGestures") { (view: GoogleMapsView, enable: Bool) in
        view.setEnabledScrollGestures(enabled: enable)
      }

      Prop("enableTiltGestures") { (view: GoogleMapsView, enable: Bool) in
        view.setEnabledTiltGestures(enabled: enable)
      }

      Prop("enableZoomGestures") { (view: GoogleMapsView, enable: Bool) in
        view.setEnabledZoomGestures(enabled: enable)
      }

      Prop("mapType") { (view: GoogleMapsView, mapType: MapType) in
        view.setMapType(mapType: mapType)
      }

      Prop("googleMapsJsonStyleString") { (view: GoogleMapsView, jsonStyleString: String) in
        view.setMapStyle(jsonStyleString: jsonStyleString)
      }

      Prop("markers") { (view: GoogleMapsView, markerObjects: [MarkerObject]) in
        view.setMarkers(markerObjects: markerObjects)
      }
      
      Prop("clusters") { (view: GoogleMapsView, clusterObjects: [ClusterObject]) in
        view.setClusters(clusterObjects: clusterObjects)
      }

      Prop("polygons") { (view: GoogleMapsView, polygonObjects: [PolygonObject]) in
        view.setPolygons(polygonObjects: polygonObjects)
      }

      Prop("polylines") { (view: GoogleMapsView, polylineObjects: [PolylineObject]) in
        view.setPolylines(polylineObjects: polylineObjects)
      }

      Prop("initialCameraPosition") { (view: GoogleMapsView, cameraPosition: CameraPosition) in
        view.setInitialCameraPosition(initialCameraPosition: cameraPosition)
      }

      Prop("circles") { (view: GoogleMapsView, circleObjects: [CircleObject]) in
        view.setCircles(circleObjects: circleObjects)
      }
      
      Prop("enableTraffic") { (view: GoogleMapsView, enable: Bool) in
        view.setEnabledTraffic(enableTraffic: enable)
      }
      
      Prop("kmls") { (view: GoogleMapsView, kmlObjects: [KMLObject]) in
        view.setKMLs(kmlObjects: kmlObjects)
      }
      
      Prop("geojsons") { (view: GoogleMapsView, geoJsonObjects: [GeoJsonObject]) in
        view.setGeoJsons(geoJsonObjects: geoJsonObjects)
      }
      
      Prop("overlays") { (view: GoogleMapsView, overlayObjects: [OverlayObject]) in
        view.setOverlays(overlayObjects: overlayObjects)
      }
    }
  }
}
